presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PhysicalResourceOptimizer.java

```diff
@@ -0,0 +1,308 @@
+        return finalExecutorCount;
+    }
+
+    private int calculateDircuOptimizedExecutorCountForShorterQueries(int numberOfExecutors)
+    {
+        int updatedExecutorCount = Math.max(numberOfExecutors, MIN_EXECUTOR_FOR_SMALL_QUERIES_DIRCU_OPTIMIZED);
+        int finalExecutorCount = Math.max(MIN_EXECUTOR_FOR_SMALL_QUERIES_DIRCU_OPTIMIZED, Math.min(MAX_EXECUTOR_FOR_SMALL_QUERIES_DIRCU_OPTIMIZED, updatedExecutorCount));
+        return finalExecutorCount;
+    }
+
+    private int calculateDircuOptimizedExecutorCountForLongerQueries(int numberOfExecutors)
+    {
+        int updatedExecutorCount = Math.max(numberOfExecutors, MIN_EXECUTOR_FOR_LARGE_QUERIES_DIRCU_OPTIMIZED);
+        int finalExecutorCount = Math.max(MIN_EXECUTOR_FOR_LARGE_QUERIES_DIRCU_OPTIMIZED, Math.min(MAX_EXECUTOR_FOR_LARGE_QUERIES_DIRCU_OPTIMIZED, updatedExecutorCount));
+        return finalExecutorCount;
+    }
+
+    private int calculateLatencyOptimizedExecutorCountForLongerQueries(int numberOfExecutors)
+    {
+        int updatedExecutorCount = Math.max(numberOfExecutors, MIN_EXECUTOR_FOR_LARGE_QUERIES_LATENCY_OPTIMIZED);
+        int finalExecutorCount = Math.max(MIN_EXECUTOR_FOR_LARGE_QUERIES_LATENCY_OPTIMIZED, Math.min(MAX_EXECUTOR_FOR_LARGE_QUERIES_LATENCY_OPTIMIZED, updatedExecutorCount));
+        return finalExecutorCount;
+    }
+    public class CollectSourceStats
+    {
+        private final Metadata metadata;
+        private final Session session;
+
+        public CollectSourceStats(Metadata metadata, Session session)
+        {
+            this.metadata = metadata;
+            this.session = session;
+        }
+
+        public AbstractMap.SimpleImmutableEntry<Double, Boolean> collectSourceStats(PlanNode root)
+        {
+            Visitor visitor = new Visitor();
+            root.accept(visitor, null);
+
+            AbstractMap.SimpleImmutableEntry<List<TableStatistics>, Boolean> sourceStatistics = visitor.getStatsForSources();
+
+            Iterator<TableStatistics> tableStatisticsIterator = sourceStatistics.getKey().iterator();
+            Double totalSourceSize = 0.0;
+            while (tableStatisticsIterator.hasNext()) {
+                TableStatistics ts = tableStatisticsIterator.next();
+                totalSourceSize = totalSourceSize + ts.getTotalSize().getValue();
+            }
+
+            return new AbstractMap.SimpleImmutableEntry(totalSourceSize, sourceStatistics.getValue());
+        }
+
+        private class Visitor
+                extends InternalPlanVisitor<Void, Void>
+        {
+            private final ImmutableList.Builder<TableStatistics> tableStatisticsBuilder = ImmutableList.builder();
+            private Boolean isSourceMissingData = false;
+
+            public AbstractMap.SimpleImmutableEntry<List<TableStatistics>, Boolean> getStatsForSources()
+            {
+                return new AbstractMap.SimpleImmutableEntry(tableStatisticsBuilder.build(), isSourceMissingData);
+            }
```

Comment:
Instead of creating `AbstractMap.SimpleImmutableEntry<List<TableStatistics>, Boolean>` object, we can just add getters to get these fields. Right?

Reply:
Updated the interface, we simply return PhysicalResourceSettings which has all the relevant info.